### ISSUE TITLE
Linux kernel crash/hung task [hv-sock proxy (vsudd) is not reachable]

### ISSUE THREAD

**Issue by rn** [docker/for-win#1994]
There have been many reports with this error message:
```
Docker hv-sock proxy (vsudd) is not reachable
```
It indicates that the Hyper-V Linux VM did not start (completely). There are several causes for this, but there are a few common root causes.

One of them is that Linux kernel crashes during boot or has does not complete the boot due to hung kernel tasks. This causes a timeout in Docker for Windows to be triggered which eventually will display the error message.

Other possible root cause are:
- Running Docker for Windows inside Windows 10 VM, aka nested virtualization. Please see [this entry in the troubleshoot](https://docs.docker.com/docker-for-windows/troubleshoot/#running-docker-for-windows-in-nested-virtualization-scenarios) for details.
- The Linux VM boot very slowly, possibly due to a high system load. Please see https://github.com/docker/for-win/issues/1995 for details.

This issue is for tracking kernel crash/hung tasks related issues.

## How to detect a Linux kernel crash/hung task:
Look at the log file (`Diagnose and Feedback` in the whale systray icon, the click on `log file`). Then look for the start of the kernel boot. It typically starts with (note there may be several in the log file):
```
[14:56:59.464][Linux          ][Info   ] Trying to connect to vsudd...
[14:57:00.166][Moby           ][Info   ] Connected
[14:57:01.236][Moby           ][Info   ] [    0.000000] Linux version 4.9.60-linuxkit-aufs (root@4a42478ffb9a) (gcc version 6.3.0 (Alpine 6.3.0) ) #1 SMP Mon Nov 6 16:00:12 UTC 2017
[14:57:01.270][Moby           ][Info   ] [    0.000000] Command line: BOOT_IMAGE=/boot/kernel console=ttyS0 page_poison=1 ntp=gateway vsyscall=emulate panic=1 root=/dev/sr0 text
[14:57:01.299][Moby           ][Info   ] [    0.000000] x86/fpu: Legacy x87 FPU detected.
[14:57:01.318][Moby           ][Info   ] [    0.000000] x86/fpu: Using 'eager' FPU context switches.
[14:57:01.335][Moby           ][Info   ] [    0.000000] e820: BIOS-provided physical RAM map:
[14:57:01.354][Moby           ][Info   ] [    0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009ffff] usable
[14:57:01.370][Moby           ][Info   ] [    0.000000] BIOS-e820: [mem 0x00000000000c0000-0x00000000000fffff] reserved
[14:57:01.384][Moby           ][Info   ] [    0.000000] BIOS-e820: [mem 0x0000000000100000-0x000000007eee9fff] usable
[14:57:01.408][Moby           ][Info   ] [    0.000000] BIOS-e820: [mem 0x000000007eeea000-0x000000007eef1fff] ACPI data
```

If you then scroll down, look out for a pattern, such as:
```
------------[ cut here ]------------
```
followed by more lines including a dump of the CPU registers etc. and another `cut here` line indicating the end of the crash dump.

**If you see this, please file a separate issue and indicate "kernel crash" in the subject**

However, there are, two know patterns:

## Context switch code crash

We have only seen this on Windows builds 14393 (RS1/Anniversary Update) and SkyLake or newer processors. SkyLake change the way the context (CPU registers etc) are saved on a context switch. The Linux kernel used in Docker for Windows tries to use this new mechanism and then there are some bad interactions with the version of Hyper-V in the 14393 build.

The crash in the logs looks something like this (omitting the common prefixes):
```
[   13.979617] ------------[ cut here ]------------
[   13.989608] WARNING: CPU: 0 PID: 0 at ./arch/x86/include/asm/fpu/internal.h:368 __switch_to+0x347/0x49d
[   13.989608] Modules linked in:
[   13.989608] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G        W       4.9.87-linuxkit-aufs #1
[   13.989608] Hardware name: Microsoft Corporation Virtual Machine/Virtual Machine, BIOS Hyper-V UEFI Release v1.0 11/26/2012
[   13.989608]  0000000000000000 ffffffff8e438de6 0000000000000000 0000000000000000
[   13.989608]  ffffffff8e0dead0 ffff8cedf9222b00 ffffffff8ee13500 0000000000000001
[   13.989608]  0000000000000000 0000000000000000 01d25a1bd3913515 ffffffff8e0200ba
[   13.989608] Call Trace:
[   13.989608] ---[ end trace 1c47b109ffec3a89 ]---
```
*Note: the `./arch/x86/include/asm/fpu/internal.h` line.*

or this:
```
[    0.000000] ------------[ cut here ]------------
[    0.000000] WARNING: CPU: 0 PID: 0 at arch/x86/kernel/fpu/xstate.c:593 fpu__init_system_xstate+0x4c1/0x8b3
[    0.000000] XSAVE consistency problem, dumping leaves
[    0.000000] Modules linked in:
[    0.000000] CPU: 0 PID: 0 Comm: swapper Not tainted 4.9.87-linuxkit-aufs #1
[    0.000000]  0000000000000000 ffffffff87438de6 ffffffff87e03e40 0000000000000000
[    0.000000]  ffffffff870dead0 000000000000000a ffffffff87e03ea0 00000000000003c0
[    0.000000]  0000000000000440 0000000000000001 000000000000008e ffffffff870deb43
[    0.000000] Call Trace:
[    0.000000]  [<ffffffff87438de6>] ? dump_stack+0x5a/0x6f
[    0.000000]  [<ffffffff870dead0>] ? __warn+0xd2/0xeb
[    0.000000]  [<ffffffff870deb43>] ? warn_slowpath_fmt+0x5a/0x76
[    0.000000]  [<ffffffff870daafa>] ? __cpuid.constprop.11+0x7/0x8
[    0.000000]  [<ffffffff87f87878>] ? fpu__init_system_xstate+0x4c1/0x8b3
[    0.000000]  [<ffffffff87f86f91>] ? fpu__init_system+0x1f8/0x2fb
[    0.000000]  [<ffffffff87f83cd4>] ? setup_arch+0xa8/0x993
[    0.000000]  [<ffffffff87f79b85>] ? start_kernel+0xa9/0x44f
[    0.000000]  [<ffffffff87f79120>] ? early_idt_handler_array+0x120/0x120
[    0.000000]  [<ffffffff87f793fd>] ? x86_64_start_kernel+0x141/0x165
[    0.000000] ---[ end trace cc2f7240d7ae6709 ]---
```

**If you see either of these in the log file, please install the latest updates. You even have to update to RS2 or newer as this seems to be a Hyper-V which has been fixed in newer version**

## Hung task in vmbus

Occasionally, we have seen a kernel thread hanging in the Linux's VMBus implementation. The VMBus implements the interaction between the Linux VM and Hyper-V on the host. If a kernel thread hangs, it most likely holds on to locks and other shared resource causes networking and other interactions with the VM (such as forwarding `docker` commands from the host to the VM) to stall.

The hung tasks can be detected in the logs by messages like:
```
[  535.510131] INFO: task kworker/0:0:3642 blocked for more than 120 seconds.
[  535.618427]       Not tainted 4.9.60-linuxkit-aufs #1
[  535.672499] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[  535.745344] kworker/0:0     D    0  3642      2 0x00000000
[  535.784769] Workqueue: hv_vmbus_con vmbus_onmessage_work
[  535.823130]  ffff8da10065d604 ffff8da0f0706840 ffff8da0f07b1480 ffff8da101face40
[  535.881933]  ffff8da09d7c6240 ffff8da101618900 ffffffff9e7f5e03 0000000000000002
[  535.936168]  ffff8da09d7c6240 ffffb47a45fdfad0 ffff8da08dcb4a28 0000000000000002
[  536.020276] Call Trace:
[  536.041663]  [<ffffffff9e7f5e03>] ? __schedule+0x364/0x465
[  536.081681]  [<ffffffff9e7f7bc9>] ? usleep_range+0x66/0x66
[  536.146439]  [<ffffffff9e7f5f82>] ? schedule+0x7e/0x87
[...]
```
or something like this:
```
[  431.858436] INFO: rcu_sched self-detected stall on CPU[  431.860041] INFO: rcu_sched detected stalls on CPUs/tasks:
[  431.860044]         0-...: (1 ticks this GP) idle=c25/140000000000001/0 softirq=7744/7744 fqs=0
[  431.860045]
[  431.860049] (detected by 1, t=12002 jiffies, g=2169, c=2168, q=2395)
[  431.860049] Task dump for CPU 0:
[  431.860050] ksoftirqd/0     R
[  431.860050]   running task        0     3      2 0x00000008 ffffffffb263269e
[  431.860054]  ffffffffb20e2bfe 0000000000000040 0000000000000100 ffffffffb27fb018
[  431.860055]  0000000000000038 0000000700000006 00000001000004a2 0000000a04208040
[  431.860057]  ffff97b380b84080 ffffffffb2e3c420 ffff97b381083ae0Call Trace:
[  431.860063]  [<ffffffffb263269e>] ? wrmsrl+0x10/0x13

[...]

[  431.860955] Task dump for CPU 0:
[  431.860955] ksoftirqd/0     R  running task        0     3      2 0x00000008
[  431.860955]  0000000000000000 ffffffffb2102408 ffffffffb2e4f980 ffffffffb2e4f780
[  431.860955]  ffffffffb219d060 0000000000000000 ffffffffb2e4f780 ffff97b381619640
[  431.860955]  ffffffffb2f3fe20 0000000000000001 ffffffffb2124979 0000000000000aa7
[  431.860955] Call Trace:
[  431.860955]  <IRQ> [  431.860955]  [<ffffffffb2102408>] ? sched_show_task+0xc8/0xd5
[  431.860955]  [<ffffffffb219d060>] ? rcu_dump_cpu_stacks+0x8e/0xa1
[  431.860955]  [<ffffffffb2124979>] ? rcu_check_callbacks+0x232/0x5e6
[  431.860955]  [<ffffffffb213534a>] ? tick_sched_do_timer+0x29/0x29
[  431.860955]  [<ffffffffb21289f3>] ? update_process_times+0x28/0x4a
[  431.860955]  [<ffffffffb2134e38>] ? tick_sched_handle+0x40/0x4c
[  431.860955]  [<ffffffffb213537e>] ? tick_sched_timer+0x34/0x58
[  431.860955]  [<ffffffffb2129255>] ? __hrtimer_run_queues+0xca/0x1a2
[  431.860955]  [<ffffffffb21296e1>] ? hrtimer_interrupt+0x64/0x11f
[  431.860955]  [<ffffffffb2633415>] ? vmbus_isr+0x70/0xb0
[  431.860955]  [<ffffffffb2034775>] ? hyperv_vector_handler+0x35/0x44
[  431.860955]  [<ffffffffb27f9ac2>] ? hyperv_callback_vector+0x82/0x90
[  431.860955]  <EOI> [  431.860955]  [<ffffffffb20415d4>] ? native_write_msr+0x6/0x1c
[  431.860955]  [<ffffffffb263269e>] ? wrmsrl+0x10/0x13
[...]
```

This issue may not just happen during boot (thus causing the `hv-sock proxy (vsudd) is not reachable` message, but also under heavy load during normal operation, after which `docker commands` will stop working (but it takes 120s or so for these messages to appear).

Unfortunately, there currently is no know fix for this issue. We have discussed this issue with the relevant Microsoft teams for the past few months and it is unclear yet, if this is an issue with Linux driver or the host side implementation or both.

If the error happens During boot, there is very little to change. If you see the error during normal operation, it may help if you change your interaction with `docker` a little as it is more likely to happen under very high load (loads of shortlived network connections from host `localhost` to a docker container, loads of docker interactions, etc).

**Comment by rn** [comment]
For reference, the following users experienced the Context switch code crash above (all on 14393). I have already closed most of the issues as we can't fix them, so this is mostly for reference, but please do report back if you still have the same issue on newer builds, either here or the sister ticket at https://github.com/docker/for-win/issues/1995):
- @dmjones500: https://github.com/docker/for-win/issues/1957
  - `E6912385-BFDF-4685-AACD-2280FC6B2844/2018-04-17_10-51-50`
- @mik-sher: https://github.com/docker/for-win/issues/1110#issue-259181410
  - `74164A4A-3009-417D-83F7-799DF43A0E8E/2017-09-20_17-11-12`
- @oskaremil: https://github.com/docker/for-win/issues/1110#issuecomment-339592283
  - `8B452E8B-650B-4A5E-A84C-005F45D175CD/2017-10-26_10-29-48`
- @joeldevries: https://github.com/docker/for-win/issues/1096#issue-257753658
  - `8D07753C-F2CC-47B6-9C9F-13F5D6D9F370/2017-09-14_10-54-08`
- @dAti: https://github.com/docker/for-win/issues/701#issue-226034297
  - `6CB81B89-A57F-420B-9288-A095769640D8/2017-05-03_17-52-36`
- @jkutil: https://github.com/docker/for-win/issues/614#issue-218607065
  - `1F0BC284-4721-450A-AF5C-7D4A5627CFE2/2017-03-31_15-11-29`
- @tomheinisch: https://github.com/docker/for-win/issues/614#issuecomment-293515744
  - `B490FB11-ADD4-4814-A71F-9ACB525F1A95/2017-04-12_10-23-58`
- @skravtsov: https://github.com/docker/for-win/issues/606#issue-216959669
  - `3D39A47D-B309-4A57-A800-D55982BE8150/2017-03-25_11-25-11`
- @strever: https://github.com/docker/for-win/issues/592#issue-214586710
  - `E2335A3B-7A2B-4DE8-81C1-8785425D7A28/2017-03-16_11-05-44`

**Comment by rn** [comment]
@AnthonyMastrean you seem to be the most recent "victim" of the hung task issue in VMBus. I noted some suggestions at the end of the bug description above on what you could try to avoid it, but sadly we don't yet have a solid fix.

For reference, here are the issues/diagnostics where we found the hung tasks issues outlined above:
- @AnthonyMastrean: https://github.com/docker/for-win/issues/1866#issuecomment-375356036
  - `EC10B2E4-09D2-4873-84AF-24452C6B5916/2018-03-19_10-10-35` has two different types of hangs
  - `EC10B2E4-09D2-4873-84AF-24452C6B5916/2018-03-19_10-14-20` ditto
  - `EC10B2E4-09D2-4873-84AF-24452C6B5916/2018-03-19_10-17-29` ditto
  - `EC10B2E4-09D2-4873-84AF-24452C6B5916/2018-04-12_11-34-10` ditto
- @flaksirus: https://github.com/docker/for-win/issues/1096#issuecomment-331930820
  - `387F6446-5B89-4E13-89AA-3128CA65EB82/2017-09-25_19-06-40`

**Comment by ArcaneEngineer** [comment]
This happened as I had insufficient disk space on C:\ to run the instance - as soon as I freed up from only 220Mb to 1.2Gb, the instance started fine.

So Docker really needs to get some idea of the size of the instance and inform user accordingly on startup attempt.

**Comment by rn** [comment]
@ArcaneEngineer I'm pretty sure that non of the hangs in this issue are related to available space.

**Comment by ArcaneEngineer** [comment]
@rn Genuinely a bit puzzled by the response here.
Spent a good amount of time getting setup with docker. Failed consistently on 6 or more startups after installation today. Quite frustrating and was about to go with something else.

Realised I had very low disk space on C - 200Mb is quite low, I'm sure you'll agree.
THE VERY MOMENT I cleared up space (and did *nothing* else differently), Docker instance started up fine. No reboots, no config changes, nothing. JUST THAT.

I come here to post the news, thinking this will help someone, and no less than one of the devs basically tells me "WTF do you know pal, you're wrong, because I'm pretty sure I know my product."

Let me ask you this: What kind of a way is this to treat newcomers to your product? Are you more interested in solving the issue, or in self-serving views of what *isn't* broken about your product?

I have no reason to mislead you.

**Comment by ArcaneEngineer** [comment]
@rn Genuinely a bit puzzled by the response here.
Spent a good amount of time getting setup with docker. Failed consistently on 6 or more startups after installation today. Quite frustrating and was about to go with something else.

Realised I had very low disk space on C - 200Mb is quite low, I'm sure you'll agree.
THE VERY MOMENT I cleared up space (and did nothing else differently), Docker instance started up fine. No reboots, no config changes, nothing. JUST THAT.

I come here to post the news, thinking this will help someone, and no less than one of the devs basically says to me, "What do you know pal, you're wrong, because I'm pretty sure I know my product."

Let me ask you this: What kind of a way is this to treat newcomers to your product? Are you more interested in solving the issue, or in self-serving views of what isn't broken about your product?

I have no reason to mislead you.

**Comment by rn** [comment]
@ArcaneEngineer I appreciate that you are reporting the result (and I'm glad that Docker for Windows is working for you now). But this issue has a specific purpose.

There were many issues related to Docker for Windows not starting (resulting in  the `vsudd not reachable` error) and I spent time going through all the supplied diagnostics and categorized the errors. There were three common root causes for the error, which covered all the reported failures. I created tracking issues for two of them. This being one of them.

So this issue is specifically to provide users with help for one of the common root causes, namely kernel crashes during boot. Being low on diskspace is likely another but a different root cause (which I hadn't seen before), but it is off topic for this issue.
